src/drivers/scsi/scsi.c

```diff
@@ -69,6 +59,105 @@ int scsi_do_cmd(struct scsi_dev *dev, struct scsi_cmd *cmd) {
 	return scsi_cmd(dev, scmd, cmd->scmd_len, cmd->scmd_obuf, cmd->scmd_olen);
 }
 
+static int scsi_send_inquiry(struct scsi_dev *dev) {
+	struct scsi_cmd cmd = scsi_cmd_template_inquiry;
+	struct scsi_data_inquiry *data;
+
+	log_debug("");
+
+	cmd.scmd_obuf = dev->scsi_data_scratchpad;
+	cmd.scmd_olen = USB_SCSI_SCRATCHPAD_LEN;
+	if (scsi_do_cmd(dev, &cmd) < 0) {
+		return -1;
+	}
+
+	data = (struct scsi_data_inquiry *) dev->scsi_data_scratchpad;
+	if ((data->dinq_devtype & SCSI_INQIRY_DEVTYPE_MASK)
+			!= SCSI_INQIRY_DEVTYPE_BLK) {
+		return -1;
+	}
+	return 0;
+}
+
+static int scsi_send_test_unit(struct scsi_dev *dev) {
+	struct scsi_cmd cmd = scsi_cmd_template_test_unit;
+
+	log_debug("");
+
+	cmd.scmd_olen = 0;
+	return scsi_do_cmd(dev, &cmd);
+}
+
+static int scsi_send_sense(struct scsi_dev *dev) {
+	struct scsi_cmd cmd = scsi_cmd_template_sense;
+	struct scsi_data_sense *data;
+	uint8_t acode;
+
+	log_debug("");
+
+	cmd.scmd_obuf = dev->scsi_data_scratchpad;
+	cmd.scmd_olen = sizeof(struct scsi_data_sense);
+	if (scsi_do_cmd(dev, &cmd) < 0) {
+		return -1;
+	}
+
+	data = (struct scsi_data_sense *) dev->scsi_data_scratchpad;
+	acode = data->dsns_additional_code;
+	if (!(acode == 0x28 || acode == 0x29)) {
+		log_error("Don't know how to recover unknown error %x", acode);
+		return -1;
+	}
+	return 0;
+}
+
+static int scsi_send_capacity10(struct scsi_dev *dev) {
+	struct scsi_cmd cmd = scsi_cmd_template_cap10;
+	struct scsi_data_cap10 *data;
+
+	log_debug("");
+
+	cmd.scmd_obuf = dev->scsi_data_scratchpad;
+	cmd.scmd_olen = sizeof(struct scsi_data_cap10);
+	if (scsi_do_cmd(dev, &cmd) < 0) {
+		return -1;
+	}
+
+	data = (struct scsi_data_cap10 *) dev->scsi_data_scratchpad;
+	dev->blk_size = be32toh(data->dc10_blklen);
+	dev->blk_n = be32toh(data->dc10_lba);
+	return 0;
+}
+
+int scsi_dev_attached(struct scsi_dev *dev) {
+	int ret;
+	int test_unit_retries = 3;
+
+	if (scsi_send_inquiry(dev)) {
+		log_debug("INQUIRY failed");
+		return -1;
+	}
+
+	/* FIXME This logic derived from the previos variant,
```

Comment:
We send sence command if inquiry failed, flash drive can be not ready yet

Reply:
Are sure about `if inquiry failed`?

Reply:
I am actually not sure, because it is not so in master. We send inquiry if test unit failed

Reply:
I will try to investigate in more detail how it should be...